copilot-verifier/src/Copilot/Verifier.hs

```diff
@@ -695,13 +744,82 @@ copilotTypeToLLVMType = loop
     loop CT.Double = L.PrimType (L.FloatType L.Double)
     loop t0@(CT.Array tp) =
       let len = fromIntegral (tylength t0) in
-      L.Array len (loop tp)
+      L.Array len (copilotTypeToLLVMTypeBool8 tp)
     loop (CT.Struct v) =
       L.Struct (map val (CT.toValues v))
 
     val :: forall t. CT.Value t -> L.Type
-    val (CT.Value tp _) = loop tp
+    val (CT.Value tp _) = copilotTypeToLLVMTypeBool8 tp
+
+-- | Like 'copilotTypeToLLVMType', except that 'CT.Bool's are assumed to be
+-- eight bits, not one bit. See @Note [How LLVM represents bool]@.
+copilotTypeToLLVMTypeBool8 ::
+  CT.Type a ->
+  L.Type
+copilotTypeToLLVMTypeBool8 CT.Bool = L.PrimType (L.Integer 8)
+copilotTypeToLLVMTypeBool8 tp = copilotTypeToLLVMType tp
 
+-- | Like 'copilotTypeToLLVMType', except that 'CT.Array's are converted to
+-- 'L.PtrTo' instead of direct 'L.Array's. See @Note [Arrays]@.
+copilotTypeToLLVMTypeArrayPtr ::
+  CT.Type a ->
+  L.Type
+copilotTypeToLLVMTypeArrayPtr (CT.Array tp) = L.PtrTo (copilotTypeToLLVMTypeBool8 tp)
+copilotTypeToLLVMTypeArrayPtr tp = copilotTypeToLLVMType tp
+
+{-
```

Comment:
Does this need to be a haddock comment for the links to work?

Reply:
I don't think Haddock would link to this no matter what comment style we used, alas. I'm just co-opting [GHC's Note style](https://www.stackbuilders.com/news/the-notes-of-ghc), which provides a convention for putting long-form comments that are frequently referenced in a central location. Unfortunately, the Notes aren't visible from the Haddocks.